Source/Tests/LinuxSyscalls/SyscallsSMCTracking.cpp

```diff
@@ -0,0 +1,326 @@
+	Size = FEXCore::AlignUp(Size, FHU::FEX_PAGE_SIZE);
+
+  {
+    FHU::ScopedSignalMaskWithUniqueLock lk(_SyscallHandler->VMATracking.Mutex);
+
+    VMATracking.ClearUnsafe(CTX, Base, Size);
+  }
+
+  if (SMCChecks != FEXCore::Config::CONFIG_SMC_NONE) {
+    FEXCore::Context::InvalidateGuestCodeRange(CTX, (uintptr_t)Base, Size);
+  }
+}
+
+void SyscallHandler::TrackMprotect(uintptr_t Base, uintptr_t Size, int Prot) {
+	Size = FEXCore::AlignUp(Size, FHU::FEX_PAGE_SIZE);
+
+  {
+    FHU::ScopedSignalMaskWithUniqueLock lk(_SyscallHandler->VMATracking.Mutex);
+
+    VMATracking.ChangeUnsafe(Base, Size, VMAProt{(Prot & PROT_WRITE) != 0});
+  }
+
+  if (SMCChecks != FEXCore::Config::CONFIG_SMC_NONE) {
+    FEXCore::Context::InvalidateGuestCodeRange(CTX, Base, Size);
+  }
+}
+
+void SyscallHandler::TrackMremap(uintptr_t OldAddress, size_t OldSize, size_t NewSize, int flags, uintptr_t NewAddress) {
+	OldSize = FEXCore::AlignUp(OldSize, FHU::FEX_PAGE_SIZE);
+	NewSize = FEXCore::AlignUp(NewSize, FHU::FEX_PAGE_SIZE);
+
+  {
+
+    FHU::ScopedSignalMaskWithUniqueLock lk(_SyscallHandler->VMATracking.Mutex);
+
+    auto OldVMA = VMATracking.LookupVMAUnsafe(OldAddress);
+
+    auto OldResource = OldVMA->second.Resource;
+    auto OldOffset = OldVMA->second.Offset + OldAddress - OldVMA->first;
+    auto OldOptions = OldVMA->second.Options;
+
+    LOGMAN_THROW_A_FMT(OldVMA != VMATracking.VMAs.end(), "VMA Tracking corruption");
+
+    if (OldSize == 0) {
+      // Mirror existing mapping
+      // must be a shared mapping
+      LOGMAN_THROW_A_FMT(OldResource != nullptr, "VMA Tracking error");
+      LOGMAN_THROW_A_FMT(OldOptions.Flags.Shared, "VMA Tracking error");
+      VMATracking.SetUnsafe(CTX, OldResource, NewAddress, OldOffset, NewSize, OldOptions);
+    } else {
+
+// MREMAP_DONTUNMAP is kernel 5.7+
+#ifdef MREMAP_DONTUNMAP
+      if (!(flags & MREMAP_DONTUNMAP))
+#endif
+      {
+        VMATracking.ClearUnsafe(CTX, OldAddress, OldSize, OldResource);
+      }
+
+      // Make anonymous mapping
+      VMATracking.SetUnsafe(CTX, OldResource, NewAddress, OldOffset, NewSize, OldOptions);
+    }
+  }
+
+  if (SMCChecks != FEXCore::Config::CONFIG_SMC_NONE) {
+    if (OldAddress != NewAddress) {
+      if (OldSize != 0) {
+        // This also handles the MREMAP_DONTUNMAP case
+        FEXCore::Context::InvalidateGuestCodeRange(CTX, OldAddress, OldSize);
+      }
+    } else {
+      // If mapping shrunk, Flush the unmapped region
+      if (OldSize > NewSize) {
+        FEXCore::Context::InvalidateGuestCodeRange(CTX, OldAddress + NewSize, OldSize - NewSize);
+      }
+    }
+  }
+}
+
+void SyscallHandler::TrackShmat(int shmid, uintptr_t Base, int shmflg) {
+  shmid_ds stat;
+
+  auto res = shmctl(shmid, IPC_STAT, &stat);
+  LOGMAN_THROW_A_FMT(res != -1, "shmctl IPC_STAT failed");
+
+  uint64_t Length = stat.shm_segsz;
+
+  {
+    FHU::ScopedSignalMaskWithUniqueLock lk(_SyscallHandler->VMATracking.Mutex);
+
+    // TODO
```

Comment:
There's a ticket for that (from the PR, `madvise (tracked in https://github.com/FEX-Emu/FEX/issues/1638, among other edge cases)`

Reply:
Could you add a ticket reference to the comment? This background easily gets lost in the future otherwise.

Reply:
Not sure i've followed up here, there is #1696. It's very unlikely it will be lost as there are several tickets for follow up, and I'm gonna be the one implementing them over the next weeks.